lang/cs/Org.Apache.REEF.Common/Telemetry/MetricTracker.cs

```diff
@@ -0,0 +1,221 @@
+﻿// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using Newtonsoft.Json;
+using Org.Apache.REEF.Utilities.Logging;
+
+namespace Org.Apache.REEF.Common.Telemetry
+{
+    /// <summary>
+    /// MetricData class maintains the current value of a single metric and keeps count of the number
+    /// of times this metric has been updated. If the metric is immutable, it keeps a record of updates.
+    /// Once the data has been processed, the records and count will reset.
+    /// </summary>
+    [JsonObject]
+    public sealed class MetricTracker : IObserver<IMetric>
+    {
+        private static readonly Logger Logger = Logger.GetLogger(typeof(MetricsData));
+
+        private bool _keepUpdateHistory;
+
+        private IDisposable _unsubscriber;
+
+        [JsonProperty]
+        private IMetric _mirror;
+
+        /// <summary>
+        /// List of the history of values this metric has held. If _keepUpdateHistory is false, only holds current value.
+        /// </summary>
+        [JsonProperty]
+        private ConcurrentQueue<MetricRecord> _records;
+        
+        /// <summary>
+        /// Number of times metric has been updated since last processed.
+        /// </summary>
+        [JsonProperty]
+        internal int ChangesSinceLastSink;
+
+        /// <summary>
+        /// Constructor for metricData
+        /// </summary>
+        /// <param name="metric"></param>
+        /// <param name="initialValue"></param>
+        internal MetricTracker(IMetric metric)
+        {
+            Subscribe(metric);
+            _mirror = metric;
```

Comment:
Subscribe should handle assigning _mirror.

Reply:
fixed.